### PR TITLE
fixed recording of sustained midi notes

### LINKED ISSUE #1700: sustain pedal not being recorded in piano roll
Windows 8.1 64bit
LMMS 1.1.0 64bit (and 32bit)

I've seen 2 other issues noted that were fixed earlier last year on this but I'm not able to record the sustain.  I assume my issue is the same and the MIDI data sent by my midi keyboard is not something LMMS recognizes.

I've logged the midi here:

 TIMESTAMP IN PORT STATUS DATA1 DATA2 CHAN NOTE EVENT  
 0005C341   2   2     B0    40    7F    1  ---  CC: Pedal (Sustain)  
 0005C436   2   2     B0    40    00    1  ---  CC: Pedal (Sustain)   

Thanks for any help you can provide.

### PR DESCRIPTION
**Pull request by serdnab**
Fixes #1700.
This recovers the behavior done by 4b340f7d5f10297bd3f7198bb86877823fe89ede, and later lost.
When a midinoteff comes from midi, and sustain pedal is pressed, the noteplayhandle pointer don't get nullified, so later, when sustain pedal is released, it can catch that pointer to emit the signal that recording of note is finished.

### CODE REVIEW

File: src/tracks/InstrumentTrack.cpp [hunk 1]
```diff
@@ -304,6 +311,28 @@ void InstrumentTrack::processInEvent( const MidiEvent& event, const MidiTime& ti
 				}
 				else
 				{
+					if (isSustainPedalPressed())
+					{
+						for (NotePlayHandle*& nph : m_notes)
+						{
+							if (nph != NULL)
+							{
+								if (nph->isReleased())
```
Comment (PhysSong) [inline src/tracks/InstrumentTrack.cpp]: I think using && is better than using many if() clauses. In if (A && B), B isn't evaluated when A is false, maybe you know it already.
So it is safe to write like if (nph != NULL && nph->isReleased()) and that will make the code cleaner.

Comment (tresf) [inline src/tracks/InstrumentTrack.cpp]: >  is safe to write like if (nph != NULL && nph->isReleased()) 

Or if you believe in such a thing... :) `if (nph && nph->isReleased())`

**Comment by PhysSong** [comment]
Should the target branch be master? #1700 is in 1.2.0 milestone.

**Comment by zonkmachine** [comment]
> Should the target branch be master? #1700 is in 1.2.0 milestone.

Yes, or we backport to stable-1.2.

@serdnab I've tested this and it works fine. It can't be tested completely because of https://github.com/LMMS/lmms/issues/3537.
As commented in the review you can cut down the number of if()'s.
Would it be rude to suggest you to have a look at #3537 too?

**Comment by serdnab** [comment]
Made the shortening of the if's.

> Would it be rude to suggest you to have a look at #3537 too?
I will take a look at it.

File: src/tracks/InstrumentTrack.cpp [hunk 2]
```diff
@@ -304,6 +311,25 @@ void InstrumentTrack::processInEvent( const MidiEvent& event, const MidiTime& ti
 				}
 				else
 				{
+					if (isSustainPedalPressed())
```
Comment (zonkmachine) [inline src/tracks/InstrumentTrack.cpp]: I think you can write this into an `if else` with the `else` above to simplify it further.
And then lines 324-326 can get one tab each.

**Comment by zonkmachine** [comment]
> I will take a look at it.

Awesome!

I restarted the Travis build. It sometimes will fail for reasons that are not related to the changes made.

**Comment by serdnab** [comment]
Addressed the remarks.
I realized that the "else" is not necessary, if not `isSustainPedalPressed()`, `m_sustainPedalPressed` is already false.

**Comment by zonkmachine** [comment]
> I realized that the "else" is not necessary, if not isSustainPedalPressed(), m_sustainPedalPressed is already false.

So, drop the test and just let there be: `else`

**Comment by serdnab** [comment]
I believe you misunderstood me.
The `if(isSustainPedalPressed())` is necessary because `m_sustainPedalPressed` can be true or false, what is not necessary is a possible "else" for the  `if(isSustainPedalPressed())` clause.

**Comment by zonkmachine** [comment]
> I believe you misunderstood me.

I believe you're right there... :)
Looking at this again.

> what is not necessary is a possible "else" for the if(isSustainPedalPressed()) clause.

This is true, but the `if( event.controllerValue() > MidiMaxControllerValue/2 )` above it still needs the `else`. You're testing a state, but this is the event that sets the state. Maybe you can show how you're thinking in code instead?

**Comment by PhysSong** [comment]
@serdnab Or you can drop `else` if you put current `else if` clause first, though current code looks better. I suggest keeping this code.

**Comment by serdnab** [comment]
I am not explaining well, then you are not understanding.
Anyway, the code I intend is the current one.

**Comment by zonkmachine** [comment]
The only difference I see is that with dropping the test you have a built in safe guard in case of an unexpected state. When we get a pedal off command when `m_sustainPedalPressed` is `false`. This may never happen and/or may not have any side effects at all. I don't know if I'm being picky here. If other developers are cool with this I vote merge.

**Comment by PhysSong** [comment]
@zonkmachine I suggest merging without more change.

**Comment by zonkmachine** [comment]
OK. Squashed and merged. Thank you so much Andrés for fixing a long standing bug!

**Comment by PhysSong** [comment]
@zonkmachine backport?

**Pull request merged**